kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -109,7 +108,18 @@ public CompletionStage<RequestFilterResult> onSaslAuthenticateRequest(short apiV
             return context.forwardRequest(header, request);
         }
         else {
-            return authenticate(request.authBytes())
+            var server = saslServer;
+            if (server == null) {
+                SaslAuthenticateResponseData failedResponse = new SaslAuthenticateResponseData()
+                        .setErrorCode(ILLEGAL_SASL_STATE.code())
+                        .setErrorMessage("Unexpected SASL request")
+                        .setAuthBytes(request.authBytes());
+                LOGGER.debug("SASL invalid state");
+                return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
+            }
+            this.saslServer = null;
```

Comment:
if we get a erroneous SaslAuthenticate down the wire (...,SaslHandshake, SaslAuthenticate, SaslAuthenticate) this will make sure the 2nd one is fails early with an ISE.